In [33]:
# Importing the libraries
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn import preprocessing
from collections import defaultdict
from surprise import SVD
from surprise import KNNWithMeans
from surprise import Dataset
from surprise import accuracy
from surprise import Reader
from surprise.model_selection import cross_validate
from surprise.model_selection import train_test_split

# Suppressing Warnings
import warnings
warnings.filterwarnings('ignore')

1. Import and warehouse data:

• Import all the given datasets and explore shape and size.

In [3]:
df0 = pd.read_csv("phone_user_review_file_1.csv",encoding='iso-8859-1')
df0

phone_url        date lang country  \
0       /cellphones/samsung-galaxy-s8/    5/2/2017   en      us   
1       /cellphones/samsung-galaxy-s8/   4/28/2017   en      us   
2       /cellphones/samsung-galaxy-s8/    5/4/2017   en      us   
3       /cellphones/samsung-galaxy-s8/    5/2/2017   en      us   
4       /cellphones/samsung-galaxy-s8/   5/11/2017   en      us   
...                                ...         ...  ...     ...   
374905          /cellphones/huawei-y3/  11/30/2015   ru      ru   
374906          /cellphones/huawei-y3/   11/9/2015   ru      ru   
374907          /cellphones/huawei-y3/   9/27/2015   ru      ru   
374908          /cellphones/huawei-y3/   9/25/2015   ru      ru   
374909          /cellphones/huawei-y3/    9/9/2015   es      es   

                  source               domain  score  score_max  \
0       Verizon Wireless  verizonwireless.com   10.0       10.0   
1            Phone Arena       phonearena.com   10.0       10.0   
2                 Amazon           amazon.com    6.0       10.0   
3                Samsung          samsung.com    9.2       10.0   
4       Verizon Wireless  verizonwireless.com    4.0       10.0   
...                  ...                  ...    ...        ...   
374905            Yandex     market.yandex.ru    8.0       10.0   
374906            Yandex     market.yandex.ru    8.0       10.0   
374907            Yandex     market.yandex.ru   10.0       10.0   
374908            Yandex     market.yandex.ru    8.0       10.0   
374909     Portable Gear      portablegear.nl    8.0       10.0   

                                                  extract       author  \
0       As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...    james0923   
2       Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3       Never disappointed. One of the reasons I've be...  Buster2020    
4       I've now found that i'm in a group of people t...   S Ate Mine   
...                                                   ...          ...   
374905  в салоне МТС купил по акцÐ...          NaN   
374906  В целом неплоÑ
ой девайс б...          NaN   
374907  В общем и целом аппарат нÐ...          NaN   
374908  Смартфон для нетребовате...          NaN   
374909                        goed toestel voor die prijs  louis van h   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
374905                                      Huawei Y3 U03  
374906                                      Huawei Y3 U03  
374907                                      Huawei Y3 U03  
374908                                      Huawei Y3 U03  
374909                                          Huawei Y3  

[374910 rows x 11 columns]

In [4]:
df1 = pd.read_csv("phone_user_review_file_2.csv",encoding='iso-8859-1')
df1

phone_url       date lang country  source  \
0                /cellphones/leagoo-lead-7/  4/15/2015   en      us  Amazon   
1                /cellphones/leagoo-lead-7/  5/23/2015   en      gb  Amazon   
2                /cellphones/leagoo-lead-7/  4/27/2015   en      gb  Amazon   
3                /cellphones/leagoo-lead-7/  4/22/2015   en      gb  Amazon   
4                /cellphones/leagoo-lead-7/  4/18/2015   en      gb  Amazon   
...                                     ...        ...  ...     ...     ...   
114920  /cellphones/motorola-moto-g-4-plus/  8/19/2016   en      in  Amazon   
114921  /cellphones/motorola-moto-g-4-plus/  8/18/2016   en      in  Amazon   
114922  /cellphones/motorola-moto-g-4-plus/  8/18/2016   en      in  Amazon   
114923  /cellphones/motorola-moto-g-4-plus/  8/18/2016   en      in  Amazon   
114924  /cellphones/motorola-moto-g-4-plus/  8/18/2016   en      in  Amazon   

              domain  score  score_max  \
0         amazon.com    2.0       10.0   
1       amazon.co.uk   10.0       10.0   
2       amazon.co.uk    8.0       10.0   
3       amazon.co.uk   10.0       10.0   
4       amazon.co.uk   10.0       10.0   
...              ...    ...        ...   
114920     amazon.in    2.0       10.0   
114921     amazon.in    2.0       10.0   
114922     amazon.in   10.0       10.0   
114923     amazon.in    8.0       10.0   
114924     amazon.in    6.0       10.0   

                                                  extract           author  \
0       The telephone headset is of poor quality , not...             luis   
1       This is my first smartphone so I have nothing ...       Mark Lavin   
2       Great phone. Battery life not great but seems ...           tracey   
3           Best 90 quid I've ever spent on a smart phone    Reuben Ingram   
4       I m happy with this phone.it s very good.thx team           viorel   
...                                                   ...              ...   
114920  It's getting problem when both sim working tog...      raj agrawal   
114921  Hopeless phone. Not at all user friendly. Woul...  Amazon Customer   
114922                                   Very nice phone.  Amazon Customer   
114923  Over all good phone . Give all the specificati...  Amazon Customer   
114924  I bought Moto G Plus last month.. Camera and m...      Venkidusami   

                                                  product  
0       Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
1       Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
2       Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
3       Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
4       Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
...                                                   ...  
114920       Motorola Moto G Plus, 4th Gen (White, 32 GB)  
114921       Motorola Moto G Plus, 4th Gen (White, 32 GB)  
114922       Motorola Moto G Plus, 4th Gen (White, 32 GB)  
114923       Motorola Moto G Plus, 4th Gen (White, 32 GB)  
114924       Motorola Moto G Plus, 4th Gen (White, 32 GB)  

[114925 rows x 11 columns]

In [5]:
df2 = pd.read_csv("phone_user_review_file_3.csv",encoding='iso-8859-1')
df2

phone_url       date lang  \
0       /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  11/7/2015   pt   
1       /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  10/2/2015   pt   
2       /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt   
3       /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt   
4       /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt   
...                                                 ...        ...  ...   
312956                /cellphones/samsung-galaxy-s-iii/  1/23/2014   de   
312957                /cellphones/samsung-galaxy-s-iii/  1/23/2014   de   
312958                /cellphones/samsung-galaxy-s-iii/  1/22/2014   ru   
312959                /cellphones/samsung-galaxy-s-iii/  1/22/2014   pt   
312960                /cellphones/samsung-galaxy-s-iii/  1/22/2014   de   

       country     source            domain  score  score_max  \
0           br  Submarino  submarino.com.br    6.0       10.0   
1           br  Submarino  submarino.com.br   10.0       10.0   
2           br  Submarino  submarino.com.br   10.0       10.0   
3           br  Submarino  submarino.com.br    8.0       10.0   
4           br    Colombo    colombo.com.br    8.0       10.0   
...        ...        ...               ...    ...        ...   
312956      de      Yopi            yopi.de   10.0       10.0   
312957      de      Yopi            yopi.de    NaN        NaN   
312958      ru     Yandex  market.yandex.ru    8.0       10.0   
312959      br  Submarino  submarino.com.br   10.0       10.0   
312960      de     Amazon         amazon.de   10.0       10.0   

                                                  extract  \
0       recomendo, eu comprei um, a um ano, e agora co...   
1       Comprei um pouco desconfiada do site e do celu...   
2       Muito bom o produto, obvio que tem versões me...   
3       Unica ressalva fica para a camera que poderia ...   
4       Rapidez e atenção na entrega. O aparelho é ...   
...                                                   ...   
312956  Ein Disney- Meisterwerk für jung und alt! Zei...   
312957  Nachdem ich mir einen neuen 3D fähigen Fernse...   
312958  По поводу аккумулятора. В...   
312959  ótimo primeira linha acho que todo mundo gost...   
312960  Bestes Smartphone was ich bisher hatte :) öaf...   

                                author  \
0                     herlington tesch   
1                  Luisa Silva Marieta   
2                                Cyrus   
3            Marcela Santa Clara Brito   
4       Claudine Maria Kuhn Walendorff   
...                                ...   
312956                         burnson   
312957                         Comflag   
312958                             NaN   
312959          magda vanessa ferreira   
312960                             NaN   

                                                  product  
0       Samsung Smartphone Samsung Galaxy S3 Slim G381...  
1       Samsung Smartphone Samsung Galaxy S3 Slim G381...  
2       Samsung Smartphone Samsung Galaxy S3 Slim G381...  
3       Samsung Smartphone Samsung Galaxy S3 Slim G381...  
4       Smartphone Samsung Galaxy S3 Slim, Dual Chip, ...  
...                                                   ...  
312956                       Samsung Galaxy Express I8730  
312957                       Samsung Galaxy Express I8730  
312958                 Samsung Galaxy S III GT-I9300 16Gb  
312959  Samsung Smartphone Samsung Galaxy S III I9300 ...  
312960                                                NaN  

[312961 rows x 11 columns]

In [6]:
df3 = pd.read_csv("phone_user_review_file_4.csv",encoding='iso-8859-1')
df3

phone_url        date lang country  \
0      /cellphones/samsung-s7262-duos-galaxy-ace/   3/11/2015   en      us   
1      /cellphones/samsung-s7262-duos-galaxy-ace/  17/11/2015   en      in   
2      /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
3      /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
4      /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
...                                           ...         ...  ...     ...   
98279               /cellphones/zte-grand-s-flex/    3/1/2016   de      de   
98280               /cellphones/zte-grand-s-flex/  31/12/2015   de      de   
98281               /cellphones/zte-grand-s-flex/  20/12/2015   de      de   
98282               /cellphones/zte-grand-s-flex/  24/11/2015   de      de   
98283               /cellphones/zte-grand-s-flex/  29/10/2015   de      de   

       source      domain  score  score_max  \
0      Amazon  amazon.com    2.0       10.0   
1      Zopper  zopper.com   10.0       10.0   
2      Amazon   amazon.in    4.0       10.0   
3      Amazon   amazon.in    6.0       10.0   
4      Amazon   amazon.in   10.0       10.0   
...       ...         ...    ...        ...   
98279  Amazon   amazon.de    2.0       10.0   
98280  Amazon   amazon.de    4.0       10.0   
98281  Amazon   amazon.de    2.0       10.0   
98282  Amazon   amazon.de   10.0       10.0   
98283  Amazon   amazon.de    4.0       10.0   

                                                 extract            author  \
0      was not conpatable with my phone as stated. I ...  Frances DeSimone   
1      Decent Functions and Easy to Operate Pros:- Th...     Expert Review   
2      Not Good Phone such price. Hang too much and v...   Amazon Customer   
3                                   not bad for features   Amazon Customer   
4                                      Excellent product               NHK   
...                                                  ...               ...   
98279  Dieses Handy macht kein Spa??. Die Wlan Verbin...      Anja Jentsch   
98280  Das ZTE Grand S Flex begeistert am Anfang durc...    Daniel Winkler   
98281  Das Smartphone h??lt in meinem Falle NICHT was...      Amazon Kunde   
98282  Dieses Smartphone habe ich letztes Jahr meinem...           Jens D.   
98283  Nach jeweils einem viertel Jahr hat das Smartp...      Amazon Kunde   

                                                 product  
0      Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
1                    Samsung Galaxy Star Pro S7262 Black  
2      Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
3      Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
4      Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
...                                                  ...  
98279  ZTE Grand S Flex Smartphone (12,7 cm (5 Zoll) ...  
98280  ZTE Grand S Flex Smartphone (12,7 cm (5 Zoll) ...  
98281  ZTE Grand S Flex Smartphone (12,7 cm (5 Zoll) ...  
98282  ZTE Grand S Flex Smartphone (12,7 cm (5 Zoll) ...  
98283  ZTE Grand S Flex Smartphone (12,7 cm (5 Zoll) ...  

[98284 rows x 11 columns]

In [7]:
df4 = pd.read_csv("phone_user_review_file_5.csv",encoding='iso-8859-1')
df4

phone_url       date lang country       source  \
0             /cellphones/karbonn-k1616/  7/13/2016   en      in   91 Mobiles   
1             /cellphones/karbonn-k1616/  7/13/2016   en      in   91 Mobiles   
2             /cellphones/karbonn-k1616/  7/13/2016   en      in   91 Mobiles   
3             /cellphones/karbonn-k1616/  4/25/2014   en      in      Naaptol   
4             /cellphones/karbonn-k1616/  4/23/2013   en      in      Naaptol   
...                                  ...        ...  ...     ...          ...   
350211  /cellphones/sony-ericsson-z710i/   8/7/2006   fr      fr  GraphMobile   
350212  /cellphones/sony-ericsson-z710i/   8/5/2006   fr      fr  GraphMobile   
350213  /cellphones/sony-ericsson-z710i/  7/19/2006   fr      fr  GraphMobile   
350214  /cellphones/sony-ericsson-z710i/   7/9/2006   fr      fr  GraphMobile   
350215  /cellphones/sony-ericsson-z710i/  6/27/2006   fr      fr  GraphMobile   

                 domain  score  score_max  \
0         91mobiles.com    2.0       10.0   
1         91mobiles.com    6.0       10.0   
2         91mobiles.com    4.0       10.0   
3           naaptol.com   10.0       10.0   
4           naaptol.com   10.0       10.0   
...                 ...    ...        ...   
350211  graphmobile.com   10.0       10.0   
350212  graphmobile.com    9.0       10.0   
350213  graphmobile.com   10.0       10.0   
350214  graphmobile.com    9.0       10.0   
350215  graphmobile.com   10.0       10.0   

                                                  extract           author  \
0       I bought 1 month before. currently speaker is ...        venkatesh   
1       I just bought one week back, I have Airtel con...           Venkat   
2       one problem in this handset opera is not worki...           krrish   
3       here Karbonn comes up with an another excellen...  BRIJESH CHAUHAN   
4       What a phone, all so on Naaptol my god 23% off...    Suraj CHAUHAN   
...                                                   ...              ...   
350211  Pour info il est sur amazon.de a 212.99€ s'i...              NaN   
350212  Habitué à samsung sony nous sort 1 jolie cla...              NaN   
350213  Pour les gens qui ne regarde pas Il fait mp3, ...              NaN   
350214                  C vrai que sans le mp3 c moyen...              NaN   
350215                  Il lui manque le mp3 quel dommage              NaN   

                      product  
0               Karbonn K1616  
1               Karbonn K1616  
2               Karbonn K1616  
3       Karbonn K1616 - Black  
4       Karbonn K1616 - Black  
...                       ...  
350211    Sony-Ericsson Z710i  
350212    Sony-Ericsson Z710i  
350213    Sony-Ericsson Z710i  
350214    Sony-Ericsson Z710i  
350215    Sony-Ericsson Z710i  

[350216 rows x 11 columns]

In [8]:
df5 = pd.read_csv("phone_user_review_file_6.csv",encoding='iso-8859-1')
df5

phone_url        date lang country  \
0       /cellphones/samsung-instinct-sph-m800/   9/16/2011   en      us   
1       /cellphones/samsung-instinct-sph-m800/   2/13/2014   en      us   
2       /cellphones/samsung-instinct-sph-m800/  12/30/2011   en      us   
3       /cellphones/samsung-instinct-sph-m800/  10/18/2008   en      us   
4       /cellphones/samsung-instinct-sph-m800/    9/6/2008   en      us   
...                                        ...         ...  ...     ...   
163832       /cellphones/alcatel-ot-club_1187/   5/12/2000   de      de   
163833       /cellphones/alcatel-ot-club_1187/   5/11/2000   de      de   
163834       /cellphones/alcatel-ot-club_1187/    5/4/2000   de      de   
163835       /cellphones/alcatel-ot-club_1187/    5/1/2000   de      de   
163836       /cellphones/alcatel-ot-club_1187/   4/25/2000   de      de   

               source             domain  score  score_max  \
0         Phone Arena     phonearena.com    8.0       10.0   
1              Amazon         amazon.com    6.0       10.0   
2         Phone Scoop     phonescoop.com    9.0       10.0   
3       HandCellPhone  handcellphone.com    4.0       10.0   
4        Reviewed.com       reviewed.com    6.0       10.0   
...               ...                ...    ...        ...   
163832           Ciao            ciao.de    2.0       10.0   
163833           Ciao            ciao.de   10.0       10.0   
163834           Ciao            ciao.de    2.0       10.0   
163835           Ciao            ciao.de    8.0       10.0   
163836           Ciao            ciao.de    2.0       10.0   

                                                  extract        author  \
0       I've had the phone for awhile and it's a prett...    ajabrams95   
1       to be clear it is not the sellers fault that t...     Stephanie   
2       Well i love this phone. i have had ton of phon...      snickers   
3       I have had my Instinct for several months now ...           A4C   
4       i have had this instinct phone for about two m...     betaBgood   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                          product  
0             Samsung Instinct HD  
1       Samsung SPH M800 Instinct  
2                   Instinct M800  
3                Samsung Instinct  
4                Samsung Instinct  
...                           ...  
163832    Alcatel Club Plus Handy  
163833    Alcatel Club Plus Handy  
163834    Alcatel Club Plus Handy  
163835    Alcatel Club Plus Handy  
163836    Alcatel Club Plus Handy  

[163837 rows x 11 columns]

• Merge all datasets onto one and explore final shape and size.

In [9]:
df=pd.concat([df0,df1,df2,df3,df4,df5],axis=0)

In [10]:
df

phone_url       date lang country  \
0          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1          /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2          /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3          /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4          /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                   ...        ...  ...     ...   
163832  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
163833  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
163834  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
163835  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
163836  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                  source               domain  score  score_max  \
0       Verizon Wireless  verizonwireless.com   10.0       10.0   
1            Phone Arena       phonearena.com   10.0       10.0   
2                 Amazon           amazon.com    6.0       10.0   
3                Samsung          samsung.com    9.2       10.0   
4       Verizon Wireless  verizonwireless.com    4.0       10.0   
...                  ...                  ...    ...        ...   
163832              Ciao              ciao.de    2.0       10.0   
163833              Ciao              ciao.de   10.0       10.0   
163834              Ciao              ciao.de    2.0       10.0   
163835              Ciao              ciao.de    8.0       10.0   
163836              Ciao              ciao.de    2.0       10.0   

                                                  extract        author  \
0       As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1       Love the phone. the phone is sleek and smooth ...     james0923   
2       Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3       Never disappointed. One of the reasons I've be...   Buster2020    
4       I've now found that i'm in a group of people t...    S Ate Mine   
...                                                   ...           ...   
163832  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
163833  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
163834  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
163835  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
163836  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                  product  
0                                       Samsung Galaxy S8  
1                                       Samsung Galaxy S8  
2       Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                           Samsung Galaxy S8 64GB (AT&T)  
4                                       Samsung Galaxy S8  
...                                                   ...  
163832                            Alcatel Club Plus Handy  
163833                            Alcatel Club Plus Handy  
163834                            Alcatel Club Plus Handy  
163835                            Alcatel Club Plus Handy  
163836                            Alcatel Club Plus Handy  

[1415133 rows x 11 columns]

In [11]:
#shape of dataframe
df.shape

(1415133, 11)

In [12]:
df.reset_index(inplace=True)

In [13]:
df.drop(columns='index',inplace=True)

In [14]:
df

phone_url       date lang country  \
0           /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
1           /cellphones/samsung-galaxy-s8/  4/28/2017   en      us   
2           /cellphones/samsung-galaxy-s8/   5/4/2017   en      us   
3           /cellphones/samsung-galaxy-s8/   5/2/2017   en      us   
4           /cellphones/samsung-galaxy-s8/  5/11/2017   en      us   
...                                    ...        ...  ...     ...   
1415128  /cellphones/alcatel-ot-club_1187/  5/12/2000   de      de   
1415129  /cellphones/alcatel-ot-club_1187/  5/11/2000   de      de   
1415130  /cellphones/alcatel-ot-club_1187/   5/4/2000   de      de   
1415131  /cellphones/alcatel-ot-club_1187/   5/1/2000   de      de   
1415132  /cellphones/alcatel-ot-club_1187/  4/25/2000   de      de   

                   source               domain  score  score_max  \
0        Verizon Wireless  verizonwireless.com   10.0       10.0   
1             Phone Arena       phonearena.com   10.0       10.0   
2                  Amazon           amazon.com    6.0       10.0   
3                 Samsung          samsung.com    9.2       10.0   
4        Verizon Wireless  verizonwireless.com    4.0       10.0   
...                   ...                  ...    ...        ...   
1415128              Ciao              ciao.de    2.0       10.0   
1415129              Ciao              ciao.de   10.0       10.0   
1415130              Ciao              ciao.de    2.0       10.0   
1415131              Ciao              ciao.de    8.0       10.0   
1415132              Ciao              ciao.de    2.0       10.0   

                                                   extract        author  \
0        As a diehard Samsung fan who has had every Sam...    CarolAnn35   
1        Love the phone. the phone is sleek and smooth ...     james0923   
2        Adequate feel. Nice heft. Processor's still sl...      R. Craig   
3        Never disappointed. One of the reasons I've be...   Buster2020    
4        I've now found that i'm in a group of people t...    S Ate Mine   
...                                                    ...           ...   
1415128  Weil mein Onkel bei ALcatel arbeitet habe ich ...    david.paul   
1415129  Hy Liebe Leserinnen und Leser!! Ich habe seit ...  Christiane14   
1415130  Jetzt hat wohl Alcatell gedacht ,sie machen wa...    michaelawr   
1415131  Ich bin seit 2 Jahren (stolzer) Besitzer eines...   claudia0815   
1415132  Was sich Alkatel hier wieder ausgedacht hat,sc...    michaelawr   

                                                   product  
0                                        Samsung Galaxy S8  
1                                        Samsung Galaxy S8  
2        Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                            Samsung Galaxy S8 64GB (AT&T)  
4                                        Samsung Galaxy S8  
...                                                    ...  
1415128                            Alcatel Club Plus Handy  
1415129                            Alcatel Club Plus Handy  
1415130                            Alcatel Club Plus Handy  
1415131                            Alcatel Club Plus Handy  
1415132                            Alcatel Club Plus Handy  

[1415133 rows x 11 columns]

In [15]:
df.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19361
author       63202
product          1
dtype: int64

In [16]:
#Round of score to the nearest integer
df['score'] = df['score'].round().astype('Int64')

In [17]:
df['score']

0          10
1          10
2           6
3           9
4           4
           ..
1415128     2
1415129    10
1415130     2
1415131     8
1415132     2
Name: score, Length: 1415133, dtype: Int64

In [18]:
#Droping irrelevant features
df.drop(columns=['phone_url','date','lang','country','source','domain','score_max','extract'],inplace=True)

In [19]:
#Impute  Missing Values 
# 1d. Check for missing values. Impute the missing values if there is any. 
# filling the null values in column 'score' and 'score_max' 

df['score'].fillna(df['score'].median(), inplace=True)
# dropping the null values in columns 'author' and 'product'
df = df.dropna()


In [20]:
df.shape

(1351931, 3)

In [21]:
#  Check for duplicate values and remove them if there is any. 
df = df.drop_duplicates()

In [22]:
# Keep only 1000000 data samples. Use random state=612
df = df.sample(n=1000000, random_state=612)

In [23]:
df

score               author  \
846575       6  carolin reichenbach   
401436       6            nande_86_   
1359243      6               Marcos   
276855       8       SidneyCrestani   
623030      10            gladys198   
...        ...                  ...   
151707       6               Carlos   
405012       8            Rachel T.   
669567      10                  aze   
175206       4              Chrissi   
270759      10             Angel D.   

                                                   product  
846575   HTC One S Smartphone (10,9 cm (4,3 Zoll) AMOLE...  
401436                                    Huawei Ascend P7  
1359243                                      Siemens CF110  
276855   Samsung Smartphone Samsung Galaxy Alpha Desblo...  
623030                                      LG G Flex D955  
...                                                    ...  
151707   LG X Power Dourado Android 6.0 Tela de 5.3 Câ...  
405012                        LG Electronics G3 black 32GB  
669567   Wiko Rainbow Lite Smartphone débloqué 4G (Ec...  
175206   Huawei P8 lite Dual-SIM Smartphone (5 Zoll (12...  
270759   Nokia Lumia 735 - Smartphone libre Windows Pho...  

[1000000 rows x 3 columns]

In [24]:
# Identify the most rated features.
#sorting on products that got highest mean score
df.groupby('product')['score'].mean().sort_values(ascending=False).head() 

product
'Sony Xperia X (F5122) – White – Dual Sim (Google Android 6.0.1, 5 Display, 2 x CORTEX A72 1.8 GHz + 4 x cortex-a53...                       10.0
Samsung Galaxy Core 2 - White                                                                                                                    10.0
Samsung Galaxy Express I437 Unlocked GSM 4G LTE Touchscreen Android Smartphone - Black                                                           10.0
Samsung Galaxy Express 2 4G LTE (G3815) - Smartphone Vodafone débloque Android (ecran 4.5", 5 Mp, 8 Go, Dual-Core 1.7 GHz, 1.5 GB RAM), Bleu    10.0
Doro PhoneEasy 622 - Móvil libre (pantalla 2.4", cámara 2 Mp, 0.04 GB), negro                                                                  10.0
Name: score, dtype: Float64

In [25]:
df.groupby('author')['score'].mean().sort_values(ascending=False).head() 

author
 efef                              10.0
STabUser                           10.0
SUAVETOCK2012                      10.0
SUAREZ_SUAREZ                      10.0
SUAREZ-DUARTE Guillermo Enrique    10.0
Name: score, dtype: Float64

In [26]:
#Identify the users with most number of reviews. 
(df['author'].value_counts()).head()



Amazon Customer    13870
Cliente Amazon      5251
Client d'Amazon     2226
Amazon Kunde        1657
Anonymous           1565
Name: author, dtype: int64

In [27]:
#ar=list(df['product'].value_counts().apply(lambda x: x>50).reset_index().index)
#dfx=df.iloc[ar,:]
#ay=list(dfx['author'].value_counts().apply(lambda x: x>50).reset_index().index)
#dfy=dfx.iloc[ay,:]

In [28]:
# extracting authors who gave greater than 50 ratings
df1 = pd.DataFrame(columns=['author', 'a_count'])
df1['author']=df['author'].value_counts().index.tolist() 
df1['a_count'] = list(df['author'].value_counts() > 50)
# get names of indexes for which count column value is False
index_names = df1[ df1['a_count'] == False ].index 
# drop these row indexes from dataFrame 
df1.drop(index_names, inplace = True) 
df1

author  a_count
0    Amazon Customer     True
1     Cliente Amazon     True
2    Client d'Amazon     True
3       Amazon Kunde     True
4          Anonymous     True
..               ...      ...
707          vincent     True
708             Hugo     True
709          Jürgen     True
710         zercan74     True
711             Joel     True

[712 rows x 2 columns]

In [29]:
# extracting product that got more than 50 ratings
df2 = pd.DataFrame(columns=['product', 'p_count'])
df2['product']=df['product'].value_counts().index.tolist() 
df2['p_count'] = list(df['product'].value_counts() > 50)
# get names of indexes for which count column value is False
index_names = df2[ df2['p_count'] == False ].index 
# drop these row indexes from dataFrame 
df2.drop(index_names, inplace = True)

In [30]:
# selecting data rows where product is having more than 50 ratings.  
df3 = df[df['product'].isin(df2['product'])] 
df3

score               author  \
846575       6  carolin reichenbach   
401436       6            nande_86_   
276855       8       SidneyCrestani   
535448      10                Rohit   
31797       10             nj_jules   
...        ...                  ...   
278934      10                  Zac   
103653      10               Febs1    
711061      10           Ольга   
1385310      4            enrico6ha   
175206       4              Chrissi   

                                                   product  
846575   HTC One S Smartphone (10,9 cm (4,3 Zoll) AMOLE...  
401436                                    Huawei Ascend P7  
276855   Samsung Smartphone Samsung Galaxy Alpha Desblo...  
535448                        Apple iPhone 5s (Gold, 16GB)  
31797    Samsung Galaxy S7 SM-G930F 32GB Smartphone (Un...  
...                                                    ...  
278934                       Sim Free Sony E5 Mobile Phone  
103653                    Samsung Galaxy S6 64GB (Verizon)  
711061        Samsung I9082 Galaxy Grand Duos (синий)  
1385310                                         Nokia 5100  
175206   Huawei P8 lite Dual-SIM Smartphone (5 Zoll (12...  

[547403 rows x 3 columns]

In [31]:
# selecting data rows from df3 where author has given more than 50 ratings.
# so that we get the data with products having more than 50 ratings and users who have given more than 50 ratings
df4 = df3[df3['author'].isin(df1['author'])]
df4

score           author  \
783134     10         Domenico   
407982      8             Luca   
875294      6           Steven   
319163     10  Amazon Customer   
88193       4  Amazon Customer   
...       ...              ...   
546110     10            Frank   
335388     10            Pablo   
889623     10              Mac   
302159      8         Fernando   
711061     10       Ольга   

                                                  product  
783134  Nokia 208 Telefono Cellulare, Micro SIM, UMTS,...  
407982  LG D855 G3 Smartphone, 16 GB, Nero Metallico [...  
875294  Samsung E1200 Sim Free Mobile Phone - Disconti...  
319163          Huawei Ascend Y550 Smartphone, 4 GB, Nero  
88193   Huawei Nexus 6P unlocked smartphone, 32GB Gold...  
...                                                   ...  
546110                    Apple iPhone 5S 16GB Space Gray  
335388  Motorola Moto G 16 GB - Smartphone libre Andro...  
889623                                        Nokia C2-05  
302159  ASUS ZenFone 2 Laser Unlocked Smartphone, 3GB ...  
711061       Samsung I9082 Galaxy Grand Duos (синий)  

[47589 rows x 3 columns]

In [32]:
# Report the shape of the final dataset.
df4.shape

(47589, 3)

In [34]:
#Build a popularity based model and recommend top 5 mobile phones.

In [35]:
#calculating the mean score for a product by grouping it.
df_pop = pd.DataFrame(df.groupby('product')['score'].mean()) 
df_pop


score
product                                                     
'Sony Xperia X (F5122) – White – Dual Sim (...      10.0
'Sony Xperia X (F5122) – rosa – Dual Sim (G...      10.0
(CUBOT) GT88 5.5" qHD 1.3GHz MTK6572 2-Core And...       8.0
(DG300 Versione Aggiornata)5'' DOOGEE VOYAGER2 ...  7.833333
(Part 4) Conclusion review of Lenovo S820, my v...       9.0
...                                                      ...
中價高階機　Dell Venue 智æ
§äºŒéƒ¨æ›² å¯...       9.0
幾可亂真　雙網雙卡白色 iPhone 4？            9.0
摩托罗拉 C168i                                      10.0
跨年前å
ˆä¸€èµ·ä¾†çœ‹ Nokia C5-03 火速試ç...       9.0
金屬 + 皮革 = 質感滿分　Wave 723 詳æ¸...       9.0

[57638 rows x 1 columns]

In [36]:
# calculating the number of ratings a product got
df_pop['rating_counts'] = pd.DataFrame(df.groupby('product')['score'].count())  

In [37]:
# 3. Recommending the 5 mobile phones based in highest mean score and highest number of ratings the product got. 
df_pop.sort_values(by=['score','rating_counts'], ascending=[False,False]).head()

score  rating_counts
product                                                                 
Motorola Smartphone Motorola Moto X Desbloquead...   10.0            147
Motorola Smartphone Motorola Moto G Dual Chip D...   10.0            141
Nokia Smartphone Nokia Lumia 520 Desbloqueado O...   10.0            140
Samsung Smartphone Galaxy Win Duos Branco Desbl...   10.0            137
Samsung Smartphone Dual Chip Samsung Galaxy SII...   10.0            134

In [38]:
df

score               author  \
846575       6  carolin reichenbach   
401436       6            nande_86_   
1359243      6               Marcos   
276855       8       SidneyCrestani   
623030      10            gladys198   
...        ...                  ...   
151707       6               Carlos   
405012       8            Rachel T.   
669567      10                  aze   
175206       4              Chrissi   
270759      10             Angel D.   

                                                   product  
846575   HTC One S Smartphone (10,9 cm (4,3 Zoll) AMOLE...  
401436                                    Huawei Ascend P7  
1359243                                      Siemens CF110  
276855   Samsung Smartphone Samsung Galaxy Alpha Desblo...  
623030                                      LG G Flex D955  
...                                                    ...  
151707   LG X Power Dourado Android 6.0 Tela de 5.3 Câ...  
405012                        LG Electronics G3 black 32GB  
669567   Wiko Rainbow Lite Smartphone débloqué 4G (Ec...  
175206   Huawei P8 lite Dual-SIM Smartphone (5 Zoll (12...  
270759   Nokia Lumia 735 - Smartphone libre Windows Pho...  

[1000000 rows x 3 columns]

In [39]:
#Build a collaborative filtering model using SVD.
# arranging columns in the order of user id,item id and rating to be fed in the svd
columns_titles = ['author','product','score']
df_c = df.reindex(columns=columns_titles)


df_c

author  \
846575   carolin reichenbach   
401436             nande_86_   
1359243               Marcos   
276855        SidneyCrestani   
623030             gladys198   
...                      ...   
151707                Carlos   
405012             Rachel T.   
669567                   aze   
175206               Chrissi   
270759              Angel D.   

                                                   product  score  
846575   HTC One S Smartphone (10,9 cm (4,3 Zoll) AMOLE...      6  
401436                                    Huawei Ascend P7      6  
1359243                                      Siemens CF110      6  
276855   Samsung Smartphone Samsung Galaxy Alpha Desblo...      8  
623030                                      LG G Flex D955     10  
...                                                    ...    ...  
151707   LG X Power Dourado Android 6.0 Tela de 5.3 Câ...      6  
405012                        LG Electronics G3 black 32GB      8  
669567   Wiko Rainbow Lite Smartphone débloqué 4G (Ec...     10  
175206   Huawei P8 lite Dual-SIM Smartphone (5 Zoll (12...      4  
270759   Nokia Lumia 735 - Smartphone libre Windows Pho...     10  

[1000000 rows x 3 columns]

In [40]:
# Keep only 5000 data samples. Use random state=612
df_c5 = df_c.sample(n=5000, random_state=612)
# 4. Build a collaborative filtering model using SVD. 

In [41]:
# 4. Build a collaborative filtering model using SVD. 
reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(df_c5,reader = reader)
data

In [42]:
#trainset = data.build_full_trainset()
# Than predict ratings for all pairs (u, i) that are NOT in the training set.
#testset = trainset.build_anti_testset()

In [43]:
# Split data to train and test
from surprise.model_selection import train_test_split
trainset, testset = train_test_split(data, test_size=.25,random_state=123)


In [44]:
algo = SVD()           #SVD(n_factors=5,biased=False)
algo.fit(trainset)

In [45]:
predictions = algo.test(testset)

In [46]:

accuracy.rmse(predictions)

RMSE: 2.4956


2.495637804524366

In [47]:
cross_validate(algo, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

Evaluating RMSE, MAE of algorithm SVD on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.5228  2.5195  2.6055  2.5405  2.5310  2.5439  0.0317  
MAE (testset)     1.9678  1.9527  2.0225  1.9775  1.9815  1.9804  0.0233  
Fit time          0.49    0.43    0.45    0.49    0.46    0.46    0.02    
Test time         0.01    0.01    0.01    0.01    0.01    0.01    0.00    


{'test_rmse': array([2.52279703, 2.51952882, 2.60549375, 2.54047226, 2.53097267]),
 'test_mae': array([1.9677712 , 1.9526602 , 2.02254035, 1.97751442, 1.98149519]),
 'fit_time': (0.4883594512939453,
  0.4298985004425049,
  0.4489166736602783,
  0.49271082878112793,
  0.4613912105560303),
 'test_time': (0.010774850845336914,
  0.010991096496582031,
  0.009796857833862305,
  0.010775327682495117,
  0.010759830474853516)}

In [48]:
def get_top_n(predictions, n=5):
    # First map the predictions to each user.
    top_n = defaultdict(list)
    for uid, iid, true_r, est, _ in predictions:
        top_n[uid].append((iid, est))

    # Then sort the predictions for each user and retrieve the k highest ones.
    for uid, user_ratings in top_n.items():
        user_ratings.sort(key=lambda x: x[1], reverse=True)
        top_n[uid] = user_ratings[:n]

    return top_n

In [49]:
# 8. Try and recommend top 5 products for test users
top_n = get_top_n(predictions, n=5)

In [50]:
top_n

defaultdict(list,
            {'lacatusu cosmin': [('iPhone 5S APPLE IPHONE 5S 32GB SILVER UNLOCKED',
               7.715598781339223)],
             'shpi-lya': [('Sony Ericsson K790i', 8.212414431470606)],
             'vijay gupt': [('Panasonic P51 - White', 8.038133333333333)],
             'Michèle': [('LG Electronics P710 Optimus L7 II Smartphone (10,9 cm (4,3 Zoll) Touchscreen, 1GHz, Dual-Core, 4GB, 768MB RAM, 8-Megapixel-Kamera, Android 4.1) metallisch-schwarz',
               7.623349482494837)],
             'deralex': [('Motorola Timeport P7389', 8.038133333333333)],
             'yaffa bey': [('3650', 8.038133333333333)],
             'mrflincki': [('Sony Ericsson W980i piano black Handy',
               8.038133333333333)],
             'Afop': [('Microsoft Nokia 6700 classic Handy (UMTS, Bluetooth, 5 MP) illuvial pink',
               8.038133333333333)],
             'andredalv': [('Alcatel One Touch Idol 3 Smartphone, 4.7", Dual SIM, Grigio Scuro [Italia]',
          

In [51]:
#Above are the top 5 predicted items and their ratings for test users.

# Print the recommended items for each user
for uid, user_ratings in top_n.items():
    print(uid, [iid for (iid, _) in user_ratings])

lacatusu cosmin ['iPhone 5S APPLE IPHONE 5S 32GB SILVER UNLOCKED']
shpi-lya ['Sony Ericsson K790i']
vijay gupt ['Panasonic P51 - White']
Michèle ['LG Electronics P710 Optimus L7 II Smartphone (10,9 cm (4,3 Zoll) Touchscreen, 1GHz, Dual-Core, 4GB, 768MB RAM, 8-Megapixel-Kamera, Android 4.1) metallisch-schwarz']
deralex ['Motorola Timeport P7389']
yaffa bey ['3650']
mrflincki ['Sony Ericsson W980i piano black Handy']
Afop ['Microsoft Nokia 6700 classic Handy (UMTS, Bluetooth, 5 MP) illuvial pink']
andredalv ['Alcatel One Touch Idol 3 Smartphone, 4.7", Dual SIM, Grigio Scuro [Italia]']
Yvan781 ['Sony Mobile Xperia Z1 Smartphone débloqué 4G (Ecran: 5 pouces - 16 Go - Android 4.1 Jelly Bean) Noir']
Client d'Amazon ["Doogee Nouveau 5.0'' DOOGEE NOVA Y100x IPS OGS 3G Smartphone Débloqué Android 5.0 Lollipop MT6582 Quad Core 1.3GHz Double SIM 8GB ROM GPS WIFI (avec Ã\x89tui de protection)- pour Orange, SFR, Bouygues, Virgin, Free ISYS, Lebara, Numericable etc (Blanc)", 'Honor 6 Plus Smart

Russn ['HTC One V']
T ['Xperia Sony Xperia SP SIM-free Android Smartphone - White (discontinued by manufacturer)']
HUH? ['HTC One M9+ Plus 32GB Gold on Silver, 5.2", GSM Unlocked International Model, No Warranty']
stang12 ['Microsoft Nokia Lumia 630 Single-SIM Smartphone (4,5 Zoll (11,4 cm) Touch-Display, 8 GB Speicher, Windows 8.1) gelb']
s3anreilly ['Samsung S7275 Galaxy Ace 3 UK SIM-Free Smartphone - White']
Tatiana Adamova ['elephone P9000 4G Smartphone MTK6755 Octa Core 2,0 GHz 4GB RAM 32GB ROM Android 6.0 5.5" Ultra Ã\x89troit 8+13Ð\x9cP Caméras Frame en Alliage d`Aluminium Empreintes Digitales Charge Rapide avec KKmoon Support']
Olegat0r ['Samsung S5830 Galaxy Ace (Black)']
Michael H. ['Sony Xperia ZL LTE Smartphone (12,7 cm (5 Zoll) Touchscreen, 1,5GHz, Quad-Core, 2GB RAM, 16GB HDD, 13 Megapixel Kamera, Android 4.1) weiÃ\x9f']
ebr-arr-01 ['Sony Ericsson S312']
Jucan ['Doro Matra - PhoneEasy 605 - T??l??phone portable - GSM - Tri-bande - gris fonc??']
olga ['BlackBerry Curve 93

In [53]:
algo_i = KNNWithMeans(k=10, sim_options={ 'user_based': False})

algo_i.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [54]:
#from surprise.model_selection import cross_validate
#cross_validate(algo_i, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)

test_pred=algo_i.test(testset)
print(accuracy.rmse(test_pred))

RMSE: 2.5125
2.512453485068154


In [55]:
from surprise.model_selection import cross_validate
cross_validate(algo_i, data, measures=['RMSE', 'MAE'], cv=5, verbose=True)


Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Computing the msd similarity matrix...
Done computing similarity matrix.
Evaluating RMSE, MAE of algorithm KNNWithMeans on 5 split(s).

                  Fold 1  Fold 2  Fold 3  Fold 4  Fold 5  Mean    Std     
RMSE (testset)    2.5998  2.6195  2.5645  2.5362  2.5230  2.5686  0.0366  
MAE (testset)     2.0098  2.0216  1.9610  1.9625  1.9631  1.9836  0.0265  
Fit time          0.87    0.53    0.43    0.43    0.45    0.54    0.17    
Test time         0.03    0.02    0.02    0.02    0.03    0.02    0.00    


{'test_rmse': array([2.59975923, 2.61954713, 2.56454752, 2.53620229, 2.52303084]),
 'test_mae': array([2.00976933, 2.02161983, 1.9610476 , 1.96245833, 1.96307775]),
 'fit_time': (0.8659303188323975,
  0.5250492095947266,
  0.43002915382385254,
  0.42806434631347656,
  0.4535372257232666),
 'test_time': (0.026445388793945312,
  0.018604755401611328,
  0.02056884765625,
  0.015673398971557617,
  0.0254666805267334)}

In [56]:
# we can now query for specific predicions
uid = 'Marcos'  # raw user id
iid = 'LG G Flex D955'  # raw item id
# get a prediction for specific users and items.
pred = algo.predict(uid, iid, verbose=True)

# run the trained model against the testset
test_pred = algo_i.test(testset)
#6. Predict score (average rating) for test users
test_pred

user: Marcos     item: LG G Flex D955 r_ui = None   est = 7.72   {'was_impossible': False}


[Prediction(uid='lacatusu cosmin', iid='iPhone 5S APPLE IPHONE 5S 32GB SILVER UNLOCKED', r_ui=10.0, est=10, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='shpi-lya', iid='Sony Ericsson K790i', r_ui=10.0, est=10, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='vijay gupt', iid='Panasonic P51 - White', r_ui=10.0, est=10, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='Michèle', iid='LG Electronics P710 Optimus L7 II Smartphone (10,9 cm (4,3 Zoll) Touchscreen, 1GHz, Dual-Core, 4GB, 768MB RAM, 8-Megapixel-Kamera, Android 4.1) metallisch-schwarz', r_ui=10.0, est=10, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='deralex', iid='Motorola Timeport P7389', r_ui=10.0, est=10, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='yaffa bey', iid='3650', r_ui=9.0, est=9.0, details={'actual_k': 1, 'was_impossible': False}),
 Prediction(uid='mrflincki', iid='Sony Ericsson W980i piano black Handy', r_ui=4

#Report your findings and inferences :

*Most popular products are	
Motorola Smartphone Motorola Moto X Desbloqueado Preto Android 4.2.2 Câmera 10MP e Frontal 2MP Memória Interna de 16GB GSM,
Motorola Smartphone Motorola Moto G Dual Chip Desbloqueado TIM Android 4.3 Tela 4.5 8GB 3G Wi-Fi Câmera 5MP - Preto,

Nokia Smartphone Nokia Lumia 520 Desbloqueado Oi Preto Windows Phone 8 Câmera 5MP 3G Wi-Fi Memória Interna 8G GPS,
Samsung Smartphone Galaxy Win Duos Branco Desbloqueado Dual Chip Câmera 5MP Processador Quad Core 1.2 Ghz Android 4.1 3G Wi- Fi e Memória 8GB,
Samsung Smartphone Dual Chip Samsung Galaxy SIII Duos Desbloqueado Claro Azul Android 4.1 3G/Wi-Fi Câmera 5MP


*products that got highest mean score -Sony Xperia X (F5122)White Dual Sim    

*Amazon Customer are the users with most number of reviews. 



*In what business scenario you should use popularity based Recommendation Systems ?

Populary based Recommendation system recommends products based on popularity and or anything which is in trend. This system is used in many business scenarios such has news businesses, hotel recommendations in a city etc. It is also used in cold start problems where we do not have knowlege about the taste of the user.

*In what business scenario you should use CF based Recommendation Systems ? 

Colleborative filtering recomment product based on the information and taste about the user. Websites like Myntra , Amazon, Nyka, Netflix, Youtube etc use colleborative filltering to recomment their products.

  
*What other possible methods can you think of which can further improve the recommendation for different users ?

Apaart from these we can combine the recommendation systems and use hybrid recommendation system, other recommendation system are Content-based, Demographic, Utility based, Knowledge based.


